rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,430 @@
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The Rust compiler generally assumes that the program that it is building is
+living in a fully isolated memory space. It leverages this knowledge to
+transform said program's memory access patterns for performance optimization
+purposes, under the assumption that said transformations will not have any
+externally observable effect other than the program running faster.
+
+Examples of such transformations include:
+
+- Caching data from RAM into CPU registers.
+- Eliminating unused loads and unobservable stores.
+- Merging neighboring stores with each other.
+- Only updating the part of an over-written struct that has actually changed.
+
+Although these optimizations are most of the time correct and useful, there are
+some situations where they are inappropriate, including but not limited to:
+
+- [Memory-mapped I/O](https://en.wikipedia.org/wiki/Memory-mapped_I/O), a common
+  low-level communication protocol between CPUs and peripherals, where hardware
+  registers masquerading as memory can be used to program peripherals by
+  accessing said registers in very specific load/store patterns.
+- [Shared memory](https://en.wikipedia.org/wiki/Shared_memory), a form of
+  inter-process communication where two programs can communicate via a common
+  memory block, and it is therefore not appropriate for Rust to assume that it
+  is aware of all memory accesses occurring inside said memory block.
+- [Cryptography](https://en.wikipedia.org/wiki/Cryptography), where it is
+  extremely important to ensure that sensitive information is erased after use,
+  and is not leaked via indirect means such as recognizable scaling patterns in
+  the time taken by a system to process attacker-crafted inputs.
+
+In such circumstances, it may be necessary to assert precise manual control on
+the memory accesses that are carried out by a Rust program in a certain memory
+region. This is the purpose of _volatile memory operations_, which allow a Rust
+programmers to generate a carefully controlled stream of hardware memory load
+and store instructions, which is guaranteed to be left untouched by the Rust
+compiler's optimizer even though surrounding Rust code will continue to be
+optimized as usual.
+
+---
+
+Volatile memory operations are exposed in the `std::volatile` module of the Rust
+standard library, or alternatively in the `core::volatile` module for the
+purpose of writing `#![no_std]` applications. They are interfaced through
+fixed-size data wrappers that are somewhat reminiscent of the API used for
+atomic operations in `std::sync::atomic`:
+
+```rust
+use std::sync::atomic::Ordering;
+use std::ptr::NonNull;
+use std::volatile::VolatileU8;
+
+unsafe fn do_volatile_things(target: NonNull<VolatileU8>) -> u8 {
+    target.store(42, Ordering::Relaxed);
+    target.load_not_atomic()
+}
+```
+
+Several specificities, however, should be apparent from the above usage example.
```

Comment:
What would you replace it with? Just removing it makes the discussion of differences start with "first of all", which feels odd.

Reply:
Maybe replace `"first of all" with "Notice that" (where notice refers to the example ?).

I'm marking this as resolved since this isn't a big issue, I just felt that the sentence was unnecessary. I'll leave it up to you if you want to reword this a bit.

Reply:
I tweaked that wording a bit in the latest version.